feat(bazel): speed up dev-turnaround by bundling types only when packaging

devversion opened the pull request:
Speeds up the dev-turnaround by only bundling types when packaging. Currently
bundling occurs for all the `ng_module` targets in devmode.

This has various positive benefits:

* Avoidance of this rather slower operation in development
* Makes APF-built packages also handle types for `ts_library` targets consistently. Simplifies the packager..
* Allows us to ensure APF entry-points have `d.ts` _always_ bundled (working with ESM
module resolution in TypeScript -- currently experimental -- https://github.com/angular/angular/issues/43886)
* Allows us to remove the secondary `package.json` files from APF (maybe APF v14? - seems
low-impact). This would clean-up the APF even more and fix resolution issues (like in Vite -- https://github.com/angular/angular/issues/45326)

devversion committed:
feat(bazel): speed up dev-turnaround by bundling types only when packaging

Speeds up the dev-turnaround by only bundling types when packaging. Currently
bundling occurs for all the `ng_module` targets in devmode.

This has various positive benefits:

* Avoidance of this rather slower operation in development
* Makes APF-built packages also handle types for `ts_library` targets consistently.
* Allows us to ensure APF entry-points have `d.ts` _always_ bundled (working with ESM
module resolution in TypeScript -- currently experimental)
* Allows us to remove the secondary `package.json` files from APF (maybe APF v14? - seems
low-impact). This would clean-up the APF even more and fix resolution issues (like in Vite)

devversion committed:
refactor: fix duplication in platform-server NPM package

The platform-server init entry-point imported code from another
entry-point using a relative import. This resulted in the code to be
bundled into the `init` entry-point as well. This has no breaking
impact but resulted in a little code duplication that we should
clean up.

devversion committed:
refactor(localize): init APF entry-point should not duplicate code

This commit addresses two issues:

* The init entry-point currenly access code from another entry-point
  using relative imports, resulting in code to be duplicated.
* The init types are now bundled as part of the ng_package APF rule.
  There is an API extractor bundling issue with global module
  augmentations.

API extractor does not properly handle module augmentation. We need to disable
dts bundling for this entry-point to ensure `$localize` remains globally accessible
for users. This is an option in the `ng_package` rule.

Note that this worked before because `localize/init` was a `ts_library` that did not
have its types bundled.

As part of this change, the `MessageId` and `TargetMessage` exports are
also made public. The localize exported functions rely on these types but
they were not exported.

Related to types, an exception is added for three private exports from the primary
entry-point so that they will show up in the API golden. These private
exports are re-exposed publicly in the init entry-point but no golden
would capture them due to the private symbol prefix. One might wonder
why the symbols are not guarded in the init golden. The reason is that
goldens never inline signatures from cross-entry-points/packages to avoid
duplication.

Lastly, the i18n integration test golden had to be updated because the
polyfills bundle increased slightly. After thorough and time-consuming
investigation, this mostly happens due to different mangle identifies
being used (the input code changed --> so the mangling determinism)

Size before this change:

```
SUCCESS: Commit undefined uncompressed runtime did NOT cross size threshold of 500 bytes or >1% (expected: 929, actual: 926).
SUCCESS: Commit undefined uncompressed main did NOT cross size threshold of 500 bytes or >1% (expected: 124544, actual: 124660).
SUCCESS: Commit undefined uncompressed polyfills did NOT cross size threshold of 500 bytes or >1% (expected: 34530, actual: 34641).
```

After:

```
SUCCESS: Commit undefined uncompressed runtime did NOT cross size threshold of 500 bytes or >1% (expected: 929, actual: 926).
SUCCESS: Commit undefined uncompressed main did NOT cross size threshold of 500 bytes or >1% (expected: 124544, actual: 124650).
FAIL: Commit undefined uncompressed polyfills exceeded expected size by 500 bytes or >1% (expected: 34530, actual: 35252).
```

Inspecting/comparing without mangling shows that the new changes would
actually result in a bundle reduction (potentially visible with
gzip/brotli):

```
➜  Desktop stat -f%z  master-nomangle.js
101357
➜  Desktop stat -f%z  with-changes-nomangle.js
101226
```

devversion committed:
fix(localize): ensure transitively loaded compiler code is tree-shakable

The localize primary entry-point (used at runtime in application code)
indirectly loads from the compiler package for computing message ids.
The compiler package has a couple of constants which cannot be DCE-ded/
tree-shaken due to side-effect reliance that is detected by Terser.

We fix these constants to be three-shakable. Note that another issue
technically would be that the compiler package has a side-effect call
for `publishFacade` (for JIT), but that invocation is marked as pure by
the Angular CLI babel optimization pipeline. So this results is no
unused code currently but is risky and should be addressed in the future.

devversion committed:
build: update systemjs aio boilerplate to work with APF v13+ `exports` field

The SystemJS examples were using an outdated version of rollup that did
not support export fields. Now with the recent changes where we removed
secondary package.json files, the rather old/somewhat outdated SystemJS
examples failed bundling since exports were not considered.

devversion committed:
test: update size golden for hello-world-ivy-compat integration test

The main bundle fell below the 500b threshold. Likely because the
global constant was fixed and is now tree-shakeable.

The actual diff in the commit is a little confusing since it makes it
seem that polyfills increased as part of this commit. This is not the
case but just a side-effect of us accumulating various changes which
are not reflected on a per-commit basis in the size golden.

The actual sizes in master were:

```
SUCCESS: Commit undefined uncompressed runtime did NOT cross size threshold of 500 bytes or >1% (expected: 1105, actual: 1102).

SUCCESS: Commit undefined uncompressed polyfills did NOT cross size threshold of 500 bytes or >1% (expected: 33846, actual: 33957).

SUCCESS: Commit undefined uncompressed main did NOT cross size threshold of 500 bytes or >1% (expected: 132392, actual: 131893).
```

Now with this change:

```
SUCCESS: Commit undefined uncompressed runtime did NOT cross size threshold of 500 bytes or >1% (expected: 1105, actual: 1102).

SUCCESS: Commit undefined uncompressed polyfills did NOT cross size threshold of 500 bytes or >1% (expected: 33846, actual: 33957).

FAIL: Commit undefined uncompressed main fell below expected size by 500 bytes or >1% (expected: 132392, actual: 131883).
```

dgp1130 commented on packages/bazel/src/types_bundle/index.ts:
```diff
@@ -0,0 +1,89 @@
+}
+
+// Entry point
+if (require.main === module) {
```
Nit: Do we need this `if` statement? Generally speaking I see this as the "top-level" Node program/binary, which should never be `import`-ed and always executed directly. If we want to import specific pieces of it for testing or other purposes, those should be factored into another `ts_library()` with no top-level side effects.

devversion replied:
Agreed. This is a leftover from the original file I refactored this from.

dgp1130 commented on packages/bazel/src/types_bundle/index.ts:
```diff
@@ -0,0 +1,89 @@
+  const extractorConfig = ExtractorConfig.prepare(options);
+  const {succeeded} = Extractor.invoke(extractorConfig);
+
+  if (succeeded) {
```
The successful case is the "interesting" part of this function, so I suggest replacing this with an early return when it didn't succeed:

```typescript
const {succeeded} = Extractor.invoke(extractorConfig);
if (!succeeded) return 1;

// ...
```

dgp1130 commented on packages/bazel/src/types_bundle/index.ts:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    // Re-write the output file.
+    fs.writeFileSync(outputExecpath, bundleOutput);
```
Nit: I recommend using `async` APIs just to avoid a ["broken window"](https://blog.codinghorror.com/the-broken-window-theory/) used to justify more synchronous stuff in the future.

devversion replied:
done

dgp1130 commented on packages/bazel/src/types_bundle/index.ts:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Strip the named AMD module for compatibility from Bazel-generated type
+ * definitions. These may end up in the generated type bundles.
+ */
```
Nit: A brief example which shows what is being stripped would be helpful to understand the regex.

dgp1130 commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -575,25 +574,27 @@ _NG_PACKAGE_ATTRS = dict(PKG_NPM_ATTRS, **{
+
+          Skipping of bundling might be desirable due to limitations in Microsoft's API extractor.
+          For example when `declare global` is used: https://github.com/microsoft/rushstack/issues/2090.
+        """,
```
Is it possible to squash this file into the initial commit? Looks like a bunch of minor changes here not too related to this commit.

dgp1130 commented on packages/compiler/src/util.ts:
```diff
@@ -134,14 +134,15 @@ declare var WorkerGlobalScope: any;
 // Check __global first, because in Node tests both __global and __window may be defined and _global
-// should be __global in that case.
-const _global: {[name: string]: any} = __global || __window || __self;
+// should be __global in that case. Note: Typeof/Instanceof checks are considered side-effects in
```
Nit: I think this comment is out of date since we've dropped `__global` and `__window`.

dgp1130 commented on packages/compiler/src/util.ts:
```diff
@@ -134,14 +134,15 @@ declare var WorkerGlobalScope: any;
+const _global: {[name: string]: any} = (/* @__PURE__ */ (
+    () => (typeof global !== 'undefined' && global) || (typeof window !== 'undefined' && window) ||
+        (typeof self !== 'undefined' && typeof WorkerGlobalScope !== 'undefined' &&
+         self instanceof WorkerGlobalScope && self))());
```
Nit: Breaking this into multiple statements would make the intent much clearer IMHO.

```typescript
const global: {[name: string]: any} = (/* @__PURE__ */ (() => {
  if (typeof global !== 'undefined') return global;
  if (typeof window !== 'undefined') return window;
  if (typeof self !== 'undefined' && typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope) return self;
  return undefined; // Or maybe `throw new Error()`?
})());
```

Same for core and localize.

devversion replied:
Totally agreed on this being more readable. If you don't feel super strongly about this, I would prefer to keep it as is to not invalidate all of the exact size analysis numbers I collected. It took quite a while to collect everything.

dgp1130 replied:
It's not that related to your change, so I'm ok with leaving this alone for now. Would be a nice cleanup to do afterwards though.

dgp1130 commented on packages/compiler/src/util.ts:
```diff
@@ -134,14 +134,15 @@ declare var WorkerGlobalScope: any;
-const _global: {[name: string]: any} = __global || __window || __self;
+// should be __global in that case. Note: Typeof/Instanceof checks are considered side-effects in
+// Terser. We explicitly mark this as side-effect free: https://github.com/terser/terser/issues/250.
+const _global: {[name: string]: any} = (/* @__PURE__ */ (
```
Do we care about this being detected as pure in JSCompiler? Should this be wrapped in `ignoreSideEffects()` as well?

devversion replied:
Good that you mention it! JSCompiler is actually able to elide this without any tag. Seems like this is just Terser behaving odd. I considered ignore-side-effects but that seemed purely for JSCompiler -- so we are good not using that here.

dgp1130 replied:
Gotcha, everything here looks pretty pure, so I think it's fine to rely on JSCompiler doing the right thing, it would be a bug on their end if it didn't work (just like it seems to be for Terser).

dgp1130 commented on packages/core/src/util/global.ts:
```diff
@@ -13,16 +13,16 @@ declare var WorkerGlobalScope: any /** TODO #9100 */;
+    () => (typeof globalThis !== 'undefined' && globalThis) ||
+        (typeof global !== 'undefined' && global) || (typeof window !== 'undefined' && window) ||
+        (typeof self !== 'undefined' && typeof WorkerGlobalScope !== 'undefined' &&
+         self instanceof WorkerGlobalScope && self))());
```
Question: Are the core and localize versions of this snippet supposed to match the one in compiler? This includes `globalThis` which should probably be preferred in compiler as well. I'm ok with not modifying behavior if that's something we don't want to chew on right now, just something which stood out to me.

Related question: Is it possible to extract this to some common utility that gets pulled into multiple packages? Last I looked into that, it seemed like `pkg_npm()` wasn't really meant to bundle files outside its own Bazel package, but I'm not sure if we even use that with `ng_package()`. Is that still the case?

devversion replied:
Yeah, let's not change the behavior here. I think you are right though. This needs to be overhauled, ideally with us sharing this code through e.g. a genrule or something. And yes, `ng_package` wraps `pkg_npm`.

dgp1130 replied:
Ok, agreed that we shouldn't change runtime behavior if we don't have to and can ignore this for now. Would be very interested to see how we can share runtime code between different packages with `pkg_npm()`, I've struggled with that a lot in the past.

dgp1130 commented on packages/bazel/src/types_bundle/BUILD.bazel:
```diff
@@ -11,14 +11,13 @@ ts_library(
 
 nodejs_binary(
-    name = "api_extractor",
+    name = "types_bundler",
```
Can we include some tests for this binary? Maybe a test which executes it as a subprocess with some test files ([example](https://github.com/dgp1130/rules_prerender/blob/main/packages/script_entry_generator/script_entry_generator_test.ts)).

Alternatively/additionally we could refactor `runMain()` and do a unit test which mocks API extractor, up to you which way is best here.

devversion replied:
Agreed, adding a test 👍

dgp1130 commented on packages/bazel/src/types_bundle/index.bzl:
```diff
@@ -0,0 +1,104 @@
+        declaration_info(output_depset),
+    ]
+
+types_bundle = rule(
```
A test for this rule would be really valuable. Can we make a simple package, extract its types with `types_bundle()` and verify the output with a golden test? Not sure if that would be brittle due to the dependency on API extractor, but maybe that would be a good thing?

devversion replied:
Yep, doing that. great idea!

dgp1130 reviewed (commented):
Thanks @devversion, lots of great stuff here! The thorough commit messages are really helpful.

I'm onboard with the overall design, mainly just missing tests. Most of the comments here are nitpicks and questions about various pieces.

devversion commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -583,6 +586,19 @@ _NG_PACKAGE_ATTRS = dict(PKG_NPM_ATTRS, **{
+    "skip_type_bundling": attr.string_list(
+        default = [],
+        doc = """
+          List of entry-points for which type bundle generation should be skipped. This might
```
Added an example. I think this would help the most. I expect this attribute to be rarely used anyway, and it going away when API extractor becomes more robust.

devversion commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -583,6 +586,19 @@ _NG_PACKAGE_ATTRS = dict(PKG_NPM_ATTRS, **{
         default = Label(_DEFAULT_ROLLUP_CONFIG_TMPL),
         allow_single_file = True,
     ),
+    "skip_type_bundling": attr.string_list(
```
I think it should be opt-out because this option is only desirable if a certain entry-point breaks with API extractor. In those cases it's then easy to add it to the exclusion list and add an obvious comment.

devversion replied:
I don't feel super strongly about this though. It probably wouldn't matter a lot. Happy to change.

devversion commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -390,13 +381,36 @@ def _ng_package_impl(ctx):
+        if entry_point in ctx.attr.skip_type_bundling:
+            if typings_entry_point.short_path != "%s/index.d.ts" % entry_point_package:
+                fail("Type bundling is explicitly disabled for `%s`, expected manual " % entry_point +
+                     "`%s/index.d.ts` to be provided." % entry_point_package)
```
good point

devversion commented on packages/bazel/src/types_bundle/index.bzl:
```diff
@@ -0,0 +1,104 @@
+
+    # Microsoft's API extractor requires a `package.json` file to be provided. We
+    # auto-generate such a file since such a file needs to exist on disk.
+    package_json = ctx.actions.declare_file(
```
good catch

devversion commented on packages/bazel/src/types_bundle/index.bzl:
```diff
@@ -0,0 +1,104 @@
+    )
+    ctx.actions.write(package_json, content = json.encode({
+        "name": "auto-generated-for-api-extractor",
+    }))
```
We need to write it to disk. Acquiring a temporary directory for building, or writing it somewhere is actually non-hermetic so this makes the most sense I think.

devversion replied:
This file will never end up being shipped, if I understand your other concern correctly.

devversion commented on packages/bazel/src/types_bundle/index.bzl:
```diff
@@ -0,0 +1,104 @@
+
+    for dep in ctx.attr.deps:
+        if DeclarationInfo in dep:
+            type_depsets.append(dep[DeclarationInfo].transitive_declarations)
```
Fair enough. I think this comes down to personal preference. I think this is a little more readable (been following this concept in most of the rules I author)

devversion commented on packages/bazel/src/types_bundle/index.bzl:
```diff
@@ -0,0 +1,104 @@
+    # Note: Using `to_list()` is expensive but we cannot get around this here as
+    # we need to find a reference to the entry-point.
+    types = depset(transitive = type_depsets)
+    types_list = types.to_list()
```
that's a great point. this would also help reducing pressure from the analysis phase.

devversion commented on packages/bazel/src/types_bundle/index.bzl:
```diff
@@ -0,0 +1,104 @@
+    # Iterate through the types and look for the entry point `File`.
+    for file in types_list:
+        if file.short_path == entry_point_short_path:
+            entry_point_file = file
```
Good point with the `break` 👍

devversion committed:
test: add test for new `types_bundle` rule from Angular bazel package

Adds a little golden test for the new `types_bundle` rule that ensures
the rule works at a general level. This rule will be useful for non-APF
ESM packages like the Angular compiler-cli (for which we also want to
bundle types to make them compatible with TypeScripts ESM type
resolution)

devversion commented:
Thank you! Addressed your feedback @dgp1130. Please have another look when you get a chance (also @josephperrott since I saw you around here 😛)

dgp1130 commented on packages/bazel/test/types_bundle/BUILD.bazel:
```diff
@@ -0,0 +1,30 @@
+
+ts_library(
+    name = "test_transitive_lib",
+    testonly = True,
```
Nit: This package might be a good use case for [`default_testonly`](https://docs.bazel.build/versions/main/be/functions.html#package.default_testonly) since it should never have production targets in it.

devversion replied:
good idea. I don't think we followed that pattern elsewhere, but it's a good practice 👍

dgp1130 commented on packages/bazel/test/types_bundle/index.ts:
```diff
@@ -0,0 +1,100 @@
+/**
```
Is this file a Git error? Looks like an unused copy of the `types_bundle` implementation. Probably should be deleted?

devversion replied:
😱 yes. I lazily copied the source folder to set up the tests haha

dgp1130 commented on packages/bazel/test/types_bundle/BUILD.bazel:
```diff
@@ -0,0 +1,30 @@
+load("@build_bazel_rules_nodejs//:index.bzl", "generated_file_test")
```
Nit: I recommend moving this commit immediately after the initial implementation (the `feat(bazel)` one) or just squashing it into that commit directly. That way tests are included as close to the code under test as possible.

dgp1130 reviewed (commented):
Looks great! Just a couple more minor things. Tests for `types_bundle` TS code still seem to be missing and I think the `types.to_list()` is still there, but otherwise pretty much ready to go.

devversion commented:
@dgp1130 Addressed your recent comments. Thank you! Regarding your review message:

* The `to_list` extraction still exists, but no longer considers transitive dependencies (like you suggested).
* I was thinking a golden test is sufficient for the types bundling. I think this covers the TS logic sufficiently as well.

dgp1130 reviewed (approved):
> * The `to_list` extraction still exists, but no longer considers transitive dependencies (like you suggested).

Ah right. I was hoping we could drop it entirely with `args.add_all()`, but that doesn't really fit what we want to do here. Limiting to direct deps should already be better than we need, so this look good.

> * I was thinking a golden test is sufficient for the types bundling. I think this covers the TS logic sufficiently as well.

That's fine with me, the golden test should give enough coverage here.

devversion commented:
I ran a [presubmit/TGP](cl/442187312) and it passes. Mostly was a little worried about the closure `Object.keys` trick in the compiler.

josephperrott reviewed (approved):
LGTM

We definitely should have the others review as well since its not just types/build changes.

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -286,11 +286,6 @@ type R3ComponentMetadataFacadeNoPropAndWhitespace = Pick<
     R3ComponentMetadataFacade,
     Exclude<Exclude<keyof R3ComponentMetadataFacade, 'preserveWhitespaces'>, 'propMetadata'>>;
 
-const USE_CLASS = Object.keys({useClass: null})[0];
```
This seems okay in principle - g3 doesn't allow JIT so this package never really gets fed to Closure, which is the only optimizer that tries to rename object keys.

I wonder if a better solution though is to try and move the `localize` -> `compiler` dependency into its own entrypoint (`@angular/compiler/i18n`), or move the JIT compiler and its facade to a separate package (which would be a breaking change in its own right).

devversion replied:
Yeah, I think both are great options. This is actually why I proposed having something like `@angular/compiler/init`, or similar (we chatted about this briefly). 

I think this would be a good follow-up. I like the non-breaking idea of having something like the `private` entry-points we have in the compiler-cli package. For now this PR should be good.

atscott reviewed (approved):
reviewed-for: fw-router

devversion commented:
Caretaker note: This is the first presubmit I ran for a PR, so please double-check if this is right. I ran a TGP with train a few days ago, and now after rebasing a normal TAP presubmit. After deflaking I forced-green and attached the `test/` link.

Additionally, before syncing this I think this CL should be incorporated: cl/443330513

atscott commented:
This PR was merged into the repository by commit 612d6e0d3d9902fae0e6350b19bb07b3b7730779.
